scripts-dev/mypy_synapse_plugin.py

```diff
@@ -92,10 +93,35 @@ def cached_function_method_signature(ctx: MethodSigContext) -> CallableType:
     arg_names.append("on_invalidate")
     arg_kinds.append(ARG_NAMED_OPT)  # Arg is an optional kwarg.
 
+    # Finally we ensure the return type is a Deferred.
+    if (
+        isinstance(signature.ret_type, Instance)
+        and signature.ret_type.type.fullname == "twisted.internet.defer.Deferred"
+    ):
+        # If it is already a Deferred, nothing to do.
+        ret_type = signature.ret_type
+    else:
+        # If a coroutine, wrap the coroutine's return type in a Deferred.
+        if (
+            isinstance(signature.ret_type, Instance)
+            and signature.ret_type.type.fullname == "typing.Coroutine"
+        ):
+            ret_arg = signature.ret_type.args[2]
+
+        # Otherwise, wrap the return type in a Deferred.
+        else:
+            ret_arg = signature.ret_type
+
+        # This should be able to use ctx.api.lookup_typeinfo, but that doesn't seem
+        # to find the correct symbol.
+        sym = ctx.api.modules["twisted.internet.defer"].names.get("Deferred")  # type: ignore[attr-defined]
```

Comment:
Which attr is unknown here---`.names`?

Reply:
`.modules`.

Reply:
That's mildly terrifying. But mypy plugins are uncharted waters!

Reply:
I expanded the comment.